### PR TITLE
In 3.4 make refutable patterns in a for comprehension an error

### PR DESCRIPTION
**Pull request by bishabosha**
supersedes #16665

Only make refutable patterns in a for comprehension an error, here we have a clear set in stone solution: put `case` before the pattern.

It is still in the air the ideal solution for pattern val definitions, see https://contributors.scala-lang.org/t/pre-sip-replace-non-sensical-unchecked-annotations/6342/85, so keep those as a warning for now

### CODE REVIEW

**Comment by carlosedp** [comment]
Do this PR supports the direct tuple assignment in a for comprehension?

**Comment by bishabosha** [comment]
> Do this PR supports the direct tuple assignment in a for comprehension?

not sure quite what you mean, but e.g. 
```scala
for (a, b) <- ((1 to 10) zip ('a' to 'z'))
yield s"$a$b"
```
will desugar to
```scala
((1 to 10) zip ('a' to 'z'))
  .map((a, b) => s"$a$b")
```
instead of 
```scala
((1 to 10) zip ('a' to 'z'))
  .withFilter({ case (a, b) => true; case _ => false })
  .map((a, b) => s"$a$b")
```

File: tests/neg/refutable-pattern-binding-messages.check [hunk 1]
```diff
@@ -38,6 +22,22 @@
    | If the narrowing is intentional, this can be communicated by adding the `case` keyword before the full pattern,
    | which will result in a filtering for expression (using `withFilter`).
    | This patch can be rewritten automatically under -rewrite -source 3.2-migration.
+-- Error: tests/neg/refutable-pattern-binding-messages.scala:5:14 ------------------------------------------------------
+5 |  val Positive(p) = 5                     // error: refutable extractor
```
Comment (sjrd) [inline tests/neg/refutable-pattern-binding-messages.check:26]: Why did these error messages move? It seems non-optimal, as the messages are not in source order anymore.

File: tests/pos/irrefutable.scala [hunk 2]
```diff
@@ -1,22 +1,11 @@
...
-//
-// A$$anonfun$f1$1.class
-// A$$anonfun$f2$1.class
-// A$$anonfun$f2$2.class
-//
-// The extra one in $f2$ is the filter.
-//
-// !!! Marking with exclamation points so maybe someday
-// this test will be finished.
+// This tests that f1 is recognized as an irrefutable pattern and f2 is not
```
Comment (sjrd) [inline tests/pos/irrefutable.scala:1]: Does it really? Will the test fail if `f2` gets suddenly considered as irrefutable (erroneously)?

Comment (bishabosha) [reply tests/pos/irrefutable.scala:1]: I guess it should be converted to a neg test

**Comment by carlosedp** [comment]
I mean, with this we could write:

```scala
  def z = for
      (a, b) <- ZIO.fromOption(Some(1, 2))
    yield (a, b)
```

Like the behavior with "-source:future" instead of assigning the result first to a variable and then unpack the tuple:

```scala
  def y = for
      x <- ZIO.fromOption(Some(1, 2))
      (a, b) = x
    yield (a, b)

```

**Comment by bishabosha** [comment]
> I mean, with this we could write:
> 
> ```scala
>   def z = for
>       (a, b) <- ZIO.fromOption(Some(1, 2))
>     yield (a, b)
> ```
> 
> Like the behavior with "-source:future" instead of assigning the result first to a variable and then unpack the tuple:
> 
> ```scala
>   def y = for
>       x <- ZIO.fromOption(Some(1, 2))
>       (a, b) = x
>     yield (a, b)
> ```

yes this PR makes for comprehensions behave like `-source:future`, but now its with `-source:3.4` (or newer)

File: tests/pos/irrefutable.scala (hunk 2 above)
Comment (bishabosha) [reply tests/pos/irrefutable.scala]: I replaced this with both a run test and a neg test

File: tests/neg/refutable-pattern-binding-messages.check (hunk 1 above)
Comment (bishabosha) [reply tests/neg/refutable-pattern-binding-messages.check:26]: all that was changed was the warning became an error, so perhaps under `-Werror` true errors print last?

**Review by sjrd — APPROVED** [review]

**Pull request merged**